flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java

```diff
@@ -250,56 +258,64 @@ public void setCurrentDatabase(String databaseName) throws DatabaseNotExistExcep
 	public Optional<CatalogTableOperation> resolveTable(String... tablePath) {
 		checkArgument(tablePath != null && tablePath.length != 0, "Table path must not be null or empty.");
 
-		List<String> defaultPath = new ArrayList<>();
-		defaultPath.add(currentCatalogName);
-		defaultPath.add(currentDatabaseName);
+		List<String> userPath = asList(tablePath);
 
-		List<String> userPath = Arrays.asList(tablePath);
-		defaultPath.addAll(userPath);
+		List<List<String>> prefixes = asList(
+			asList(currentCatalogName, currentDatabaseName),
+			singletonList(currentCatalogName),
+			emptyList()
+		);
 
-		Optional<CatalogTableOperation> inDefaultPath = lookupPath(defaultPath);
-
-		if (inDefaultPath.isPresent()) {
-			return inDefaultPath;
-		} else {
-			return lookupPath(userPath);
+		for (List<String> prefix : prefixes) {
+			Optional<CatalogTableOperation> potentialTable = lookupPath(prefix, userPath);
+			if (potentialTable.isPresent()) {
+				return potentialTable;
+			}
 		}
+
+		return Optional.empty();
 	}
 
-	private Optional<CatalogTableOperation> lookupPath(List<String> path) {
+	private Optional<CatalogTableOperation> lookupPath(List<String> prefix, List<String> userPath) {
 		try {
-			Optional<TableSchema> potentialTable = lookupCatalogTable(path);
+			List<String> path = new ArrayList<>(prefix);
+			path.addAll(userPath);
+
+			Optional<CatalogTableOperation> potentialTable = lookupCatalogTable(path);
 
 			if (!potentialTable.isPresent()) {
 				potentialTable = lookupExternalTable(path);
 			}
-			return potentialTable.map(schema -> new CatalogTableOperation(path, schema));
+			return potentialTable;
 		} catch (TableNotExistException e) {
 			return Optional.empty();
 		}
 	}
 
-	private Optional<TableSchema> lookupCatalogTable(List<String> path) throws TableNotExistException {
-		if (path.size() >= 3) {
+	private Optional<CatalogTableOperation> lookupCatalogTable(List<String> path) throws TableNotExistException {
```

Comment:
the signature has been confusing. Returning an `Optional` implies that it will be empty if the operation doesn't find the target, but the method is also throwing `TableNotExistException` when the target table is not found.

I'd suggest try-catch `TableNotExistException` in `lookupCatalogTable()` and, in case of it, return an empty `Optional`. And remove the try-catch in `lookupPath()`

Reply:
Good catch!